core/src/serde/ser.rs

```diff
@@ -0,0 +1,486 @@
+        self.serialize_i128(i128::from(value))
+    }
+
+    fn serialize_i128(self, value: i128) -> Result<Self::Ok, Self::Error> {
+        Ok(Self::Ok::Integer(value))
+    }
+
+    #[inline]
+    fn serialize_u8(self, value: u8) -> Result<Self::Ok, Self::Error> {
+        self.serialize_i128(value.into())
+    }
+
+    #[inline]
+    fn serialize_u16(self, value: u16) -> Result<Self::Ok, Self::Error> {
+        self.serialize_i128(value.into())
+    }
+
+    #[inline]
+    fn serialize_u32(self, value: u32) -> Result<Self::Ok, Self::Error> {
+        self.serialize_i128(value.into())
+    }
+
+    #[inline]
+    fn serialize_u64(self, value: u64) -> Result<Self::Ok, Self::Error> {
+        self.serialize_i128(value.into())
+    }
+
+    #[inline]
+    fn serialize_f32(self, value: f32) -> Result<Self::Ok, Self::Error> {
+        self.serialize_f64(f64::from(value))
+    }
+
+    #[inline]
+    fn serialize_f64(self, value: f64) -> Result<Self::Ok, Self::Error> {
+        Ok(Self::Ok::Float(value))
+    }
+
+    #[inline]
+    fn serialize_char(self, value: char) -> Result<Self::Ok, Self::Error> {
+        self.serialize_str(&value.to_string())
+    }
+
+    #[inline]
+    fn serialize_str(self, value: &str) -> Result<Self::Ok, Self::Error> {
+        Ok(Self::Ok::String(value.to_owned()))
+    }
+
+    fn serialize_bytes(self, value: &[u8]) -> Result<Self::Ok, Self::Error> {
+        Ok(Self::Ok::Bytes(value.to_vec()))
+    }
+
+    #[inline]
+    fn serialize_unit(self) -> Result<Self::Ok, Self::Error> {
+        Err(ser::Error::custom("Unit is not supported"))
+    }
+
+    #[inline]
+    fn serialize_unit_struct(self, _name: &'static str) -> Result<Self::Ok, Self::Error> {
+        Err(ser::Error::custom("Unit structs are not supported"))
+    }
+
+    #[inline]
+    fn serialize_unit_variant(
+        self,
+        _name: &'static str,
+        _variant_index: u32,
+        variant: &'static str,
+    ) -> Result<Self::Ok, Self::Error> {
+        self.serialize_str(variant)
+    }
+
+    #[inline]
+    fn serialize_newtype_struct<T: ?Sized>(
+        self,
+        name: &'static str,
+        value: &T,
+    ) -> Result<Self::Ok, Self::Error>
+    where
+        T: ser::Serialize,
+    {
+        let ipld = value.serialize(self);
+        if name == CID_SERDE_PRIVATE_IDENTIFIER {
+            if let Ok(Ipld::Bytes(bytes)) = ipld {
+                let cid = Cid::try_from(bytes)
+                    .map_err(|err| ser::Error::custom(format!("Invalid CID: {}", err)))?;
+                return Ok(Self::Ok::Link(cid));
+            }
+        }
+        ipld
+    }
+
+    fn serialize_newtype_variant<T: ?Sized>(
+        self,
+        _name: &'static str,
+        _variant_index: u32,
+        variant: &'static str,
+        value: &T,
+    ) -> Result<Self::Ok, Self::Error>
+    where
+        T: ser::Serialize,
+    {
+        let values = BTreeMap::from([(variant.to_owned(), value.serialize(self)?)]);
```

Comment:
uber-nit: insert will be faster (this will get collected into a temporary array otherwise).

Reply:
I hoped that it would be the reverse and that the compiler could optimize that one even better.

Reply:
Allocations (side effects) can make optimizations a bit difficult, but I haven't tested.